README.md
# Fetch Politely

[![Build Status](https://travis-ci.org/voxpelli/node-fetch-politely.svg?branch=master)](https://travis-ci.org/voxpelli/node-fetch-politely)
[![dependencies Status](https://david-dm.org/voxpelli/node-fetch-politely/status.svg)](https://david-dm.org/voxpelli/node-fetch-politely)
[![Known Vulnerabilities](https://snyk.io/test/github/voxpelli/node-fetch-politely/badge.svg?targetFile=package.json)](https://snyk.io/test/github/voxpelli/node-fetch-politely?targetFile=package.json)
[![js-semistandard-style](https://img.shields.io/badge/code%20style-semistandard-brightgreen.svg?style=flat)](https://github.com/Flet/semistandard)


A library for ensuring polite outgoing HTTP requests that respect robots.txt and aren't made too close to each other

## Installation

```bash
npm install fetch-politely --save
```

## Usage

Simple:

```javascript
const fetchInstance = new FetchPolitely((err, url, message) => {
  if (err) { return; }

  // The URL has been cleared for fetching – the hostname isn't throttled and robots.txt doesn't ban it
}, {
  // Robots.txt checking requires specification of a User Agent as Robots.txt can contain User Agent specific rules
  // See http://en.wikipedia.org/wiki/User_agent for more info in format
  userAgent: 'Your-Application-Name/your.app.version (http://example.com/optional/full/app/url)',
});

// When a slot has been reserved the callback sent in the constructor will be called
fetchInstance.requestSlot('http://foo.example.org/interesting/content/');
```

## FetchPolitely()

`var fetchInstance = new FetchPolitely(callback, [options]);`

### Parameters

* **callback** – `(err, url, message, [content]) => {};`, called for each succesful request slot

### Options

* **throttleDuration** – for how long in milliseconds to throttle requests to each `hostname`. Defaults to `10` seconds.
* **returnContent** – whether to fetch and return the content with the callback when a URL has received a request slot. Defaults to `false`.
* **logger** – a [Bunyan](https://github.com/trentm/node-bunyan) compatible logger library. Defaults to [bunyan-duckling](https://github.com/bloglovin/node-bunyan-duckling) which uses `console.log()`/`.error()`.
* **lookup** – an object or class that keeps track of throttled hosts and queued URL:s. Defaults to `PoliteLookup`.
* **lookupOptions** – an object that defines extra lookup options.
* **allowed** – a function that checks whether a URL is allowed to be fetched. Defaults to `PoliteRobot.allowed()`.
* **robotCache** – a cache method used by `PoliteRobot` to cache fetched `robots.txt`. Defaults to wrapped [lru-cache](https://www.npmjs.com/package/lru-cache).
* **robotCacheLimit** – a limit of the number of items to keep in the default lru-cache of `PoliteRobot`.
* **robotPool** – an HTTP agent to use for the request-library of `PoliteRobot`.
* **userAgent** – _required_ by `PoliteRobot` and `options.returnContent`. The [User Agent](http://en.wikipedia.org/wiki/User_agent) to use for HTTP requests.

### Methods

* **requestSlot** – tries to reserve a request slot for a URL. Returns a Promise that will be resolved or rejected when the request has been made.

### Static

* **FetchPolitely.PoliteError** – a very polite error object used for eg. informing about denied URL:s
* **FetchPolitely.PoliteLookup** – defines the interface for keeping track of throttled hosts and queued URL:s
* **FetchPolitely.PolitePGLookup** – alternative lookup that uses PostgreSQL as the backend
* **FetchPolitely.PoliteRobot** – checks whether URL:s are allowed to be fetched according to [Robots.txt](http://en.wikipedia.org/wiki/Robots_exclusion_standard).

## fetchInstance.requestSlot()

`fetchInstance.requestSlot(url, [message], [options]);`

### Parameters

* **url** – the URL to reserve a request slot for
* **message** – a JSON-encodeable optional message containing eg. instructions for the `FetchPolitely` callback.

### Options

* **allow** – if set to `true` the URL will always be allowd and not be sent to the `allowed` function.
* **allowDuplicates** – if set to `false` no more than one item of every `url` + `message` combination will be queued.

## PoliteLookup

The simplest of simple implementations for keeping track of throttled hosts and queued URL:s. Handles it all in-memory. Same interface can be used to build a database backend for this though.

## PolitePGLookup

A PostgreSQL + [Knex](http://knexjs.org/)-driven lookup that throttles hosts and queues URL using database tables.

Use by setting up the tables in `pglookup.sql` and include by setting the `FetchPolitely` options to:

```javascript
{
  lookup: FetchPolitely.PolitePGLookup,
  lookupOptions: {
    knex: knexInstance
  }
}
```

Pull Requests are welcome if someone wants to pull out the Knex-dependency. Most projects where this has been used with Postgres has been using Knex so it got used here as well.

### lookupOptions

* **knex** – *required* – the database connection to use, provided through a [Knex](http://knexjs.org/) object.
* **purgeWindow** – the minimum interval in milliseconds between two host purges. Defaults to `500` ms.
* **concurrentReleases** – how many parallell database lookups to perform to check for released URL:s. Defaults to `2`.
* **releasesPerBatch** – how many URL:s to fetch in each database lookup. Defaults to `5`.
* **onlyDeduplicateMessages** – bool that if set will only deduplicate URL:s with the same message when deduplicating. Defaults to `false`.

## Lint / Test

`npm test`


CHANGELOG.md
# Changelog

## 2.0.0-0 (2018-07-28)

* **Breaking:** Now requires Node.js 8.x or higher

## 1.1.0 (2016-11-04)

* **Feature:** PolitePGLookup now has a `onlyDeduplicateMessages` option to instruct it to deduplicate things differently

## 1.0.1 (2016-11-04)

* **Logging:** By default now doesn't log any debug-level log items

## 1.0.0 (2016-11-04)

* **Breaking:** Now requires at least Node.js 4.x
* **Dependencies:** Dropped the Promise polyfill
* **Dependencies:** Moved away from full lodash 2.x, mostly to ES6 methods
* **Dependencies:** Updated LRU Cache
* **Dev dependencies:** Moved away from a Grunt-based setup to a npm script, semistandard-based one

## 0.7.0 (2015-04-21)


### Features

* **robot:** improved pool handling ([af48a1a9](http://github.com/voxpelli/node-fetch-politely/commit/af48a1a95e07c3333e7238e351fb832ea486f098))


## 0.6.2 (2015-04-16)


### Bug Fixes

* **main:**
  * don’t return any value in requestSlot() ([9d55308e](http://github.com/voxpelli/node-fetch-politely/commit/9d55308e4d7bce0110d094d6191c981f7e9b3692))
  * should wait for queueing ([d2d177df](http://github.com/voxpelli/node-fetch-politely/commit/d2d177df5b77575fa2b3447051cebfc145ade1ff))
* **robot:** use synchronous resolving ([e1aba02a](http://github.com/voxpelli/node-fetch-politely/commit/e1aba02aa48d636b2d3501079a48746813ca322c))


## 0.6.1 (2015-04-16)


### Bug Fixes

* **main:**
  * don’t return any value in requestSlot() ([9d55308e](http://github.com/voxpelli/node-fetch-politely/commit/9d55308e4d7bce0110d094d6191c981f7e9b3692))
  * should wait for queueing ([d2d177df](http://github.com/voxpelli/node-fetch-politely/commit/d2d177df5b77575fa2b3447051cebfc145ade1ff))
* **robot:** use synchronous resolving ([e1aba02a](http://github.com/voxpelli/node-fetch-politely/commit/e1aba02aa48d636b2d3501079a48746813ca322c))


## 0.6.0 (2015-04-10)


### Bug Fixes

* **main:** set maxSockets to Infinity for requests ([89975bab](http://github.com/voxpelli/node-fetch-politely/commit/89975bab998d92c83c9e6ae84d6a976a7b544b5e))


### Features

* **robot:** set a fixed timeout limit ([f1ebacd6](http://github.com/voxpelli/node-fetch-politely/commit/f1ebacd6f15cc1ba1ed60b09533ca894fc2ccccf))


## 0.5.0 (2015-04-08)


### Features

* **lookup:** individual hostname throttle levels ([522f7a5a](http://github.com/voxpelli/node-fetch-politely/commit/522f7a5a14dde7f56841d280f9600cb6da45de3d))


## 0.4.0 (2015-04-07)


### Bug Fixes

* **logging:**
  * improved flood check interval ([3f77945a](http://github.com/voxpelli/node-fetch-politely/commit/3f77945ad3ca0c6eeee62ee599807aa76645524d))
  * lower warning threshold ([72851872](http://github.com/voxpelli/node-fetch-politely/commit/7285187222812877954c4f24bddc2d31b56ddec8))
  * properly filter flooded hostnames ([6dded0e2](http://github.com/voxpelli/node-fetch-politely/commit/6dded0e2a181c4bebbb6bb16275873e8ae6ec58a))


### Features

* **robot:** cache limit option ([daeb9033](http://github.com/voxpelli/node-fetch-politely/commit/daeb90333c7fb60d9340a8b2f18e3c3afa9ffa07))


## 0.3.1 (2015-04-01)


### Bug Fixes

* **logging:** default to ”debug”-level logging ([060954c6](http://github.com/voxpelli/node-fetch-politely/commit/060954c6574a07fd41c0f0aff42ffdd8132a475c))


## 0.3.0 (2015-04-01)


### Bug Fixes

* **dependencies:** updated Promise to 6.x.x ([1627d83a](http://github.com/voxpelli/node-fetch-politely/commit/1627d83a38f58edeae1ecec3cf9faa10452510bb))


### Features

* **logging:**
  * warn on large throttle queues ([4ece841e](http://github.com/voxpelli/node-fetch-politely/commit/4ece841ed63f4a6aee09dfb58207c5b1f0fc713f))
  * moved to Bunyan-compatible logger ([28211718](http://github.com/voxpelli/node-fetch-politely/commit/2821171833fe8c2df1962a58fc95bb3a7ca17075))


### Breaking Changes

* New option, ”logger”, for specifying which logger to use, the old ”log” option has been removed
 ([28211718](http://github.com/voxpelli/node-fetch-politely/commit/2821171833fe8c2df1962a58fc95bb3a7ca17075))


## 0.2.0 (2015-03-28)


### Bug Fixes

* **logging:** removed an erroneous whitespace ([79bdcb64](http://github.com/voxpelli/node-fetch-politely/commit/79bdcb640cce7f3d6fc5eac1d7ff234b0595769d))
* **lookup:**
  * only serialise when needed ([1abdb8b9](http://github.com/voxpelli/node-fetch-politely/commit/1abdb8b9b0b4d42b63685af36a2b233fc09a666e))
  * ensure messages are serializable ([cfb91b4b](http://github.com/voxpelli/node-fetch-politely/commit/cfb91b4b35033c2b444f88697a589526fdcf6afe))
* **main:** return possible errors in slot-reserve ([05abc938](http://github.com/voxpelli/node-fetch-politely/commit/05abc93830fa0b74703077409392dc680874a7bd))


### Features

* **lookup:** added a postgres driven lookup ([d68a2ad4](http://github.com/voxpelli/node-fetch-politely/commit/d68a2ad472255f3b4b50ddfdf5fa1ea6051ec76a))
* **main:**
  * can create instance of external lookup ([a025e514](http://github.com/voxpelli/node-fetch-politely/commit/a025e514e28cb5088871bc71717aea41b22c6b61))
  * added a debug options ([099ff456](http://github.com/voxpelli/node-fetch-politely/commit/099ff456fdeb44d807570f0cb43ab6a04cdb2b1b))
  * add a closing method ([6a34156a](http://github.com/voxpelli/node-fetch-politely/commit/6a34156a9254291dcf26ff6f332a7bbad4466deb))
  * option to disable allow-checks ([893ad429](http://github.com/voxpelli/node-fetch-politely/commit/893ad429eaf444510cc45e3e9e87d1e2e8c4ac9c))


lib/main.js
// @ts-check
/// <reference types="node" />

'use strict';

const assert = require('assert');

const noop = function () {};

const bunyanAdaptor = require('bunyan-adaptor')({
  verbose: noop
});

const PoliteError = require('./error');

/**
 * @typedef {function(string): Promise<boolean>} allowedCallback
 */

/**
 * @typedef {object} FetchPolitelyStructure
 * @property {object} PoliteLookup
 */

class FetchPolitely {
  /**
   * Creates a FetchPolitely instance.
   * @param {function(Error|null, string, string|Object<string,any>, [string]): void} callback
   * @param {Object} options
   * @param {string} options.userAgent
   * @param {false|allowedCallback} [options.allowed]
   * @param {Object} [options.logger]
   * @param {Object} [options.lookup]
   * @param {Object} [options.lookupOptions]
   * @param {boolean} [options.returnContent]
   * @param {Object} [options.robotCache]
   * @param {number} [options.robotCacheLimit]
   * @param {Object} [options.robotPool]
   * @param {number} [options.throttleDuration]
   */
  constructor (callback, options) {
    options = Object.assign({
      throttleDuration: 10 * 1000, // 10 seconds
      returnContent: false,
      logger: bunyanAdaptor,
      lookup: undefined,
      lookupOptions: undefined,
      allowed: undefined
    }, options);

    assert(callback, 'No callback defined for succesful slot requests');

    this.log = options.logger.debug.bind(options.logger);

    if (options.returnContent) {
      assert(options.userAgent, 'You need to define a user-agent');

      this._request = require('request').defaults({
        pool: {maxSockets: Infinity},
        headers: {
          'user-agent': options.userAgent
        }
      });
    }

    if (!options.lookup || !options.lookup.reserveSlot) {
      const LookupConstructor = options.lookup || FetchPolitely.PoliteLookup;
      options.lookup = new LookupConstructor(Object.assign({
        throttleDuration: options.throttleDuration,
        log: this.log,
        logger: options.logger
      }, options.lookupOptions || {}));
    }

    if (options.allowed === false) {
      this._allowedCallback = () => Promise.resolve(true);
    } else if (!options.allowed) {
      const robot = new FetchPolitely.PoliteRobot({
        cache: options.robotCache,
        cacheLimit: options.robotCacheLimit,
        pool: options.robotPool,
        userAgent: options.userAgent,
        log: this.log
      });
      this._allowedCallback = robot.allowed.bind(robot);
    } else {
      /** @type {allowedCallback} */
      this._allowedCallback = options.allowed;
    }

    this.options = options;
    this.logger = options.logger;
    this.callback = (err, url, message) => {
      if (options.returnContent && !err) {
        this._request(url, (err, res, result) => {
          if (!err && res.statusCode > 299) {
            err = new PoliteError('Unsuccessful fetch', 'statuscode');
            err.res = res;
          }

          if (err) {
            result = undefined;
          }

          callback.call(this, err, url, message, result);
        });
      } else {
        setImmediate(() => {
          callback.call(this, err, url, message);
        });
      }
    };

    options.lookup.releasedFromQueue(this.callback.bind(this));
  }

  /**
   * @param {string} url
   * @param {string|Object<string,any>} [message]
   * @param {Object} [options]
   * @param {boolean} [options.allow]
   * @param {boolean} [options.allowDuplicates]
   * @returns {Promise<void>} Will be resolved or rejected when the request has been made.
   */
  requestSlot (url, message, options = {}) {
    const result = options.allow === true ? Promise.resolve(true) : this._allowed(url);

    return result
      .then(allowed => {
        if (!allowed) {
          const err = new PoliteError('Not allowed to fetch that resource', 'unallowed');
          this.callback(err, url, message);
          return;
        }

        return this._reserveSlot(url).then(reserved => {
          if (reserved) {
            this.callback(null, url, message);
          } else {
            return this._queueForLater(url, message, {
              allowDuplicates: options.allowDuplicates
            });
          }
        });
      })
      .then(() => {
        // We don't want to return anything
      });
  }

  close () {
    // FIXME: Readd this if needed
    // if (this.options.lookup.close) { this.options.lookup.close(); }
    // if (this._allowedCallback.close) { this._allowedCallback.close(); }
  }

  _allowed (url) {
    return this._allowedCallback(url);
  }

  _reserveSlot () {
    return this.options.lookup.reserveSlot.apply(this.options.lookup, arguments);
  }

  _queueForLater () {
    return this.options.lookup.queueForLater.apply(this.options.lookup, arguments);
  }
}

FetchPolitely.PoliteError = PoliteError;
FetchPolitely.PoliteLookup = require('./lookup');
FetchPolitely.PolitePGLookup = require('./pglookup');
FetchPolitely.PoliteRobot = require('./robot');

module.exports = FetchPolitely;


lib/fetch.js
/*jslint node: true */
/* global -Promise */

'use strict';

var _ = require('lodash');
var Promise = require('promise');

var assert = require('assert');

var PoliteError = require('./error');
var PoliteLookup = require('./lookup');
var PoliteRobot = require('./robot');

var FetchPolitely = function (callback, options) {
  var robot;

  options = _.extend({
    throttleDuration: 10 * 1000, // 10 seconds
    returnContent: false,
    log: console.log.bind(console),
    debug: function () {},
    lookup: undefined,
    lookupOptions: undefined,
    allowed: undefined,
  }, options);

  assert(callback, 'No callback defined for succesful slot requests');

  if (options.returnContent) {
    assert(options.userAgent, 'You need to define a user-agent');

    this._request = require('request').defaults({
      headers: {
        'user-agent': options.userAgent
      },
    });
  }

  if (!options.lookup || !options.lookup.reserveSlot) {
    options.lookup = new (options.lookup || PoliteLookup)(_.extend({
      throttleDuration: options.throttleDuration,
      log: options.log,
      debug: options.debug,
    }, options.lookupOptions || {}));
  }

  if (options.allowed === false) {
    options.allowed = function () {
      return Promise.resolve(true);
    };
  } else if (!options.allowed) {
    robot = new PoliteRobot({
      cache : options.robotCache,
      userAgent : options.userAgent,
      log: options.log,
    });
    options.allowed = robot.allowed.bind(robot);
  }

  this.options = options;
  this.log = options.log;
  this.debug = options.debug;
  this.callback = function (err, url, message) {
    var that = this;

    if (options.returnContent && !err) {
      this._request(url, function (err, res, result) {
        if (!err && res.statusCode > 299) {
          err = new PoliteError('Unsuccessful fetch', 'statuscode');
          err.res = res;
        }

        if (err) {
          result = undefined;
        }

        callback.call(that, err, url, message, result);
      });
    } else {
      setImmediate(function () {
        callback.call(that, err, url, message);
      });
    }
  };

  //TODO: Add an option for fetching file content before sending it to callback – we will have the dependencies for that installed anyway due to Robots.txt

  options.lookup.releasedFromQueue(this.callback.bind(this));
};

FetchPolitely.prototype.requestSlot = function (url, message, options) {
  var that = this;
  var result;

  options = options || {};

  result = options.allow === true ? Promise.resolve(true) : this._allowed(url);

  return result.then(function (allowed) {
      var err;

      if (!allowed) {
        err = new PoliteError('Not allowed to fetch that resource', 'unallowed');
        that.callback(err, url, message);
        return;
      }

      return that._reserveSlot(url).then(function (reserved) {
        if (reserved) {
          that.callback(null, url, message);
        } else {
          that._queueForLater(url, message, {
            allowDuplicates: options.allowDuplicates
          });
        }
      });
    });
};

FetchPolitely.prototype.close = function () {
  if (this.options.lookup.close) { this.options.lookup.close(); }
  if (this.options.allowed.close) { this.options.allowed.close(); }
};

FetchPolitely.prototype._allowed = function (url) {
  return this.options.allowed(url);
};

FetchPolitely.prototype._reserveSlot = function () {
  return this.options.lookup.reserveSlot.apply(this.options.lookup, arguments);
};

FetchPolitely.prototype._queueForLater = function () {
  return this.options.lookup.queueForLater.apply(this.options.lookup, arguments);
};

FetchPolitely.PoliteError = PoliteError;
FetchPolitely.PoliteLookup = PoliteLookup;
FetchPolitely.PoliteRobot = PoliteRobot;

module.exports = FetchPolitely;


lib/robot.js
// @ts-check
/// <reference types="node" />

'use strict';

const assert = require('assert');
const request = require('request');
const robots = require('robots');
const urlModule = require('url');

const quickCache = function (options) {
  const cache = require('lru-cache')(options);

  return (key, value) => {
    if (!value) {
      return cache.get(key);
    } else {
      cache.set(key, value);
    }
  };
};

class PoliteRobot {
  /**
   * @param {Object} options
   * @param {function} options.log
   * @param {string} options.userAgent
   * @param {Object} [options.cache]
   * @param {number} [options.cacheLimit]
   * @param {Object} [options.pool] A pool for requests
   */
  constructor (options) {
    if (!options.cache) {
      options.cache = quickCache({
        max: options.cacheLimit || 500,
        maxAge: 1000 * 60 * 60 * 24
      });
    }

    assert(options.userAgent, 'Robots.txt lookups needs a user-agent to match against');

    this._request = request.defaults({
      timeout: 8000,
      pool: options.pool || false,
      headers: {
        'user-agent': options.userAgent
      }
    });
    this.options = options;
    this.log = options.log;
  }

  _cache (key, value) {
    return this.options.cache(key, value);
  }

  _fetch (url) {
    const robotUrl = urlModule.resolve(url, '/robots.txt');
    const robotContent = this._cache(robotUrl);

    if (robotContent !== undefined) {
      this.log('Found cache for', robotUrl);
      return robotContent;
    }

    this.log('Fetching', robotUrl);

    const result = new Promise((resolve, reject) => {
      this._request(robotUrl, (err, res, result) => {
        if (err) {
          return reject(err);
        }
        if ([401, 403].indexOf(res.statusCode) !== -1) {
          // Everything is forbidden
          result = false;
        } else if (res.statusCode > 399) {
          // Everything is allowed
          result = true;
        }
        resolve(result);
      });
    });

    // Make all requests for the same Robots file use the same Promise!
    this._cache(robotUrl, result);

    return result;
  }

  allowed (url) {
    return this._fetch(url).then(robotContent => {
      if (robotContent === false) {
        return false;
      } else if (robotContent === true) {
        return true;
      } else {
        // @ts-ignore See https://github.com/ekalinin/robots.js/pull/31
        const parser = new robots.RobotsParser();
        parser.parse(robotContent.split(/\r\n|\r|\n/));
        return parser.canFetchSync(this.options.userAgent, urlModule.parse(url).pathname);
        // TODO: Make use of crawldelay as well!
      }
    }).then(access => {
      this.log('We', access ? 'are' : 'are NOT', 'allowed to fetch', urlModule.parse(url).hostname + '' + urlModule.parse(url).pathname);
      return access;
    });
  }
}

module.exports = PoliteRobot;


example/example.js
// @ts-check
/// <reference types="node" />

'use strict';

const start = Date.now();

const FetchPolitely = require('../');
const polite = new FetchPolitely((err, url, message, result) => {
  console.log('Callback', Math.round((Date.now() - start) / 1000), 'seconds after start:', err, url, message, result ? result.length : undefined);
  if (err) {
    console.log(err.stack);
  }
}, {
  throttleDuration: 5000,
  userAgent: 'Fetch-Politely/dev',
  returnContent: true,
  logger: require('bunyan-adaptor')(),
  lookupOptions: {
    throttleDurationHost: hostname => Promise.resolve(hostname === 'example.com' ? 500 : undefined)
  }
});

polite.requestSlot('http://example.com/');
polite.requestSlot('http://example.com/');
polite.requestSlot('http://example.com/');
polite.requestSlot('http://example.com/');
polite.requestSlot('http://example.com/');
polite.requestSlot('http://example.com/');
polite.requestSlot('http://example.com/');
polite.requestSlot('http://example.com/');

polite.requestSlot('http://voxpelli.com/');
polite.requestSlot(
  'http://voxpelli.com/',
  undefined,
  { allow: false }
);
polite.requestSlot(
  'http://voxpelli.com/',
  { purpose: 'extract-metadata' }, // Could be anything useful to process the fetched page
  { allowDuplicates: false }
);
polite.requestSlot(
  'http://voxpelli.com/',
  { purpose: 'extract-metadata' }, // Could be anything useful to process the fetched page
  { allowDuplicates: false }
);

// Unallowed fetch
polite.requestSlot('http://google.se/search/123');

// Trigger flood warnings
// for (var i = 0; i < 1000; i++) {
//   polite.requestSlot('http://example.com/' + i);
// }


lib/pglookup.js
// @ts-check
/// <reference types="node" />

'use strict';

const VError = require('verror');

const assert = require('assert');
const urlModule = require('url');

class PolitePGLookup {
  /**
   * @param {Object} options
   * @param {function} options.log
   * @param {Object} options.logger
   * @param {Object} options.knex
   * @param {number} [options.concurrentReleases]
   * @param {boolean} [options.onlyDeduplicateMessages]
   * @param {number} [options.purgeWindow]
   * @param {number} [options.releasesPerBatch]
   * @param {number} [options.throttleDuration]
   */
  constructor (options) {
    options = Object.assign({
      throttleDuration: 10 * 1000,
      concurrentReleases: 2,
      releasesPerBatch: 5,
      purgeWindow: 500
    }, options);

    assert(options.knex, 'PolitePGLookup requires the Knex option to work.');

    this.options = options;
    this.log = options.log;
    this.logger = options.logger;
    this.knex = options.knex;
    this.onlyDeduplicateMessages = options.onlyDeduplicateMessages;
    this.concurrentReleases = 0;
  }
  _purgeUnthrottledHosts () {
    if (this.stopped || (this.lastPurge && (Date.now() - this.lastPurge < this.options.purgeWindow))) {
      return Promise.resolve();
    }

    this.logger.debug('Purging expired hosts');
    const knex = this.knex;

    this.lastPurge = Date.now();

    return knex('polite_hosts')
      .where('added', '<', knex.raw('CURRENT_TIMESTAMP - interval \'' + Math.round(this.options.throttleDuration / 1000) + ' seconds\''))
      .delete();
  }

  _checkForReleasables (retry) {
    const knex = this.knex;

    if (!retry) {
      this.concurrentReleases += 1;
    }

    const basicQuery = knex('polite_queue')
      .groupBy('polite_queue.hostname')
      .limit(this.options.releasesPerBatch);

    return this._purgeUnthrottledHosts()
      .then(() => {
        // Reserve a host (Making a raw query as Knex otherwise for some reason won't return the hostname)
        return this.stopped ? false : knex.transaction(trx => {
          // We need to ensure that no one else reserves the same thing
          return this.stopped ? false : trx.raw('LOCK TABLE polite_hosts')
            .then(() => {
              return this.stopped ? false : trx.raw(knex('polite_hosts')
                .returning('polite_hosts.hostname')
                .insert(
                  basicQuery
                    .clone()
                    .select(
                      'polite_queue.hostname',
                      knex.raw('CURRENT_TIMESTAMP AS added')
                    )
                    .leftJoin('polite_hosts', 'polite_queue.hostname', 'polite_hosts.hostname')
                    .whereNull('polite_hosts.hostname')
                    .orderByRaw('MIN(polite_queue.added) ASC')
                ).toString());
            });
        });
      })
      .catch(err => {
        this.concurrentReleases -= 1;
        throw new VError(err, 'Failed reserve slots for queued URL:s');
      })
      .then(result => {
        if (this.stopped || result.rowCount === 0) {
          this.concurrentReleases -= 1;
          return 0;
        }

        const hostnames = result.rows.map(row => row.hostname);

        const groupQuery = basicQuery.select('hostname', knex.raw('MIN(polite_queue.added) AS added'))
          .whereIn('hostname', hostnames)
          .orderBy('added', 'asc');

        // Fetch the first URL that was queued for the host
        knex.raw(
          'SELECT DISTINCT ON (polite_queue.hostname) id, url, messages FROM polite_queue ' +
          'INNER JOIN (' + groupQuery.toString() + ') AS grouped ' +
          'ON grouped.hostname = polite_queue.hostname AND grouped.added = polite_queue.added'
        )
        .then(result => {
          if (!result) {
            return Promise.reject(new VError('Couldn\'t find URL'));
          } else if (this.stopped) {
            return [];
          }

          const ids = result.rows.map(row => row.id);

            // Delete the URL from the queue
          return knex('polite_queue')
              .delete()
              .whereIn('id', ids)
              .then(() => result.rows);
        })
          .then(rows => {
            // Invoke all of the callbacks
            rows.forEach(row => {
              let messages = row.messages || [];

              messages = messages.length ? messages : [null];

              messages.forEach(message => {
                setImmediate(() => {
                  this.releaseCallback(null, row.url, message);
                });
              });
            });
          })
          .catch(err => {
            this.log('An error occured while releasing URL for hostnames, ', hostnames, '–', err.message);
          })
          .then(() => {
            this.concurrentReleases -= 1;
            this._watchForReleasables();
          });

        return result.rowCount;
      });
  }

  _watchForReleasables () {
    if (this.stopped) {
      this.logger.debug('Not going to watch due to stopping');
      return;
    }

    if (this.releaseTimer) {
      clearTimeout(this.releaseTimer);
      this.releaseTimer = undefined;
    }

    this._checkForReleasables()
      .catch(err => {
        this.log('Error when watching for released URL:s:', err.message, err.stack);
        return false;
      })
      .then(foundUrls => {
        if (foundUrls >= this.options.releasesPerBatch) {
          if (this.concurrentReleases < this.options.concurrentReleases) {
            this.logger.debug('Found', foundUrls, 'URL:s. Adding new release check:', this.concurrentReleases + 1);
            this._watchForReleasables();
          } else {
            this.logger.debug('Reached max concurrent releases of', this.options.concurrentReleases);

            if (this.releaseTimer) {
              clearTimeout(this.releaseTimer);
              this.releaseTimer = undefined;
            }
          }
        } else if (!this.releaseTimer) {
          const delay = Math.round(this.options.throttleDuration / 2 * Math.random());
          this.logger.debug('Retrying in', delay, 'ms');
          this.releaseTimer = setTimeout(() => { this._watchForReleasables(); }, delay);
        }
      });
  }

  _reserveSlot (hostname) {
    const knex = this.knex;

    return knex('polite_hosts')
      .insert(
        knex()
          .select(
            knex.raw('? AS hostname', hostname),
            knex.raw('CURRENT_TIMESTAMP AS added')
          )
          .whereNotExists(
            knex('polite_hosts').select('*').where('hostname', hostname)
          )
      )
      .then(result => result.rowCount !== 0) // Did we manage to reserve it?
      .catch(err => {
        if (parseInt(err.code, 10) === 23505) {
          // Rejected as a duplicate, someone else managed to reserve it before us
          return false;
        }

        throw new VError(err, 'Failed to add row');
      });
  }

  _upsertUniqueUrlToQueue (url, message) {
    const knex = this.knex;

    let arrayConversion, arrayAppending;

    const update = {
      updated: knex.fn.now()
    };

    if (message && !this.onlyDeduplicateMessages) {
      arrayConversion = knex()
        .as('array_conversion')
        .select(knex.raw('json_array_elements(polite_queue.messages) AS messages'));

      arrayAppending = knex()
        .from(arrayConversion)
        .select(knex.raw('array_to_json(array_agg(array_conversion.messages) || ?::json)', [JSON.stringify(message)]));

      update.messages = arrayAppending;
    }

    let query = knex('polite_queue')
      .update(update)
      .where('url', url);

    if (this.onlyDeduplicateMessages) {
      query = message ? query.whereRaw('messages::TEXT = ?', [JSON.stringify([message])]) : query.whereNull('messages');
    } else {
      query = query.whereNotNull('noduplicate');
    }

    return query
      .then(affectedRows => affectedRows ? Promise.resolve() : this._addUrlToQueue(url, message, true))
      .catch(err => {
        if (parseInt(err.code, 10) === 23505) {
          // Rejected as a duplicate, someone else managed to add it before us – try again
          return this._upsertUniqueUrlToQueue(url, message);
        }

        throw new VError(err, 'Failed to upsert unique URL to queue');
      });
  }

  _addUrlToQueue (url, message, noduplicate) {
    const knex = this.knex;
    const hostname = urlModule.parse(url).hostname;

    return knex('polite_queue').insert({
      url: url,
      messages: message ? JSON.stringify([message]) : null,
      hostname: hostname,
      updated: knex.fn.now(),
      added: knex.fn.now(),
      noduplicate: noduplicate && !this.onlyDeduplicateMessages ? true : null
    })
      .catch(err => Promise.reject(new VError(err, 'Failed to add URL to queue')));
  }

  close () {
    this.stopped = true;

    if (this.releaseTimer) {
      clearTimeout(this.releaseTimer);
      this.releaseTimer = undefined;
    }
  }

  releasedFromQueue (callback) {
    this.releaseCallback = callback;
    this._watchForReleasables();
  }

  queueForLater (url, message, options) {
    let result;

    options = Object.assign({
      allowDuplicates: undefined
    }, options || {});

    if (options.allowDuplicates !== false) {
      result = this._addUrlToQueue(url, message);
    } else {
      result = this._upsertUniqueUrlToQueue(url, message);
    }

    return result;
  }

  reserveSlot (url) {
    const hostname = urlModule.parse(url).hostname;

    return this
      ._purgeUnthrottledHosts()
      .then(() => this._reserveSlot(hostname))
      .catch(err => Promise.reject(new VError(err, 'Database error on reservation')));
  }
}

module.exports = PolitePGLookup;
